File: workdir/src/TestDriver.py
#! jython
# -*- coding: utf-8; -*-

from contextlib import closing
from java.io import BufferedReader, File, InputStreamReader
from java.lang import Class, ClassNotFoundException, System
from java.net import URL, URLClassLoader
from java.sql import SQLException
from java.util import Properties
import jarray
import os.path
import sys
import traceback

# ロギングの環境設定ファイル (※ HSQLDB-1.8.0.10 では使用しない)
#System.setProperty("java.util.logging.config.file", "logging.properties")

def add_classpath():
    """`-J-cp`オプションが存在しない処理系への対処"""
    libdir = os.path.join(os.path.dirname(__file__), "..", "lib")
    urls = jarray.array([
        File(os.path.join(libdir, "odb.jar")).toURI().toURL(),
        File(os.path.join(libdir, "hsqldb.jar")).toURI().toURL()
    ], URL)
    cl = URLClassLoader(urls)
    return cl

info = Properties()
info.setProperty("user", "sa")
info.setProperty("password", "")

try:
    with closing(add_classpath()) as cl, \
        closing(Class.forName("com.k650250.odb.ODBFile", True, cl)
            .open("sample.odb")) as odbFile, \
        closing(Class.forName("org.hsqldb.jdbcDriver", True, cl)
            .newInstance().connect(odbFile.toUrl(), info)) as con, \
        closing(con.createStatement()) as st:

        System.out.println(u"[更新前のデータ一覧]")
        #sql = 'SELECT * FROM "t_sample"'  # テーブル
        #sql = 'SELECT * FROM "v_sample"'  # ビュー
        #sql = odbFile.getQuery("q_sample").toString()          # クエリー (問合せ文中にクエリー名が含まれない場合)
        sql = odbFile.getQuery("q_SAMPLE").expand().toString()  # クエリー (問合せ文中にクエリー名が含まれる場合)
        with closing(st.executeQuery(sql)) as records:
            while records.next():
                System.out.println(u"{}\t{}".format(records.getString("key"), records.getString("value")))

        # データ挿入
        sql = 'INSERT INTO "t_sample"("value") VALUES(?)'
        with closing(con.prepareStatement(sql)) as prep:
            System.out.println(u"追加データを入力してください。")
            prep.setString(1, BufferedReader(InputStreamReader(
                eval("System.in"),
                System.getProperty("native.encoding"))).readLine())
            prep.executeUpdate()

        System.out.println(u"[更新後のデータ一覧]")
        sql = 'SELECT * FROM "t_sample"'
        with closing(st.executeQuery(sql)) as records:
            while records.next():
                System.out.println(u"{}\t{}".format(records.getString("key"), records.getString("value")))

except ClassNotFoundException as e:
    e.printStackTrace()
    System.err.println(u"org.hsqldb.jdbcDriver が見つかりません。")
except SQLException as e:
    e.printStackTrace()
    System.err.println(u"考えられる原因:")
    System.err.println(u"- hsqldb.jar が正常に読み込まれていない")
    System.err.println(u"- このデータベースを別のプロセスが使用中")
    System.err.println(u"- 問合せ文中にクエリー名が含まれる")
except Exception as e:
    traceback.print_exc(file=sys.stderr)

File: workdir/src/TestDriver.clj
;; -*- coding: utf-8; -*-

(import '[java.io BufferedReader InputStreamReader]
        '[java.net URL URLClassLoader]
        '[java.sql SQLException]
        '[java.util Properties])
(require '[clojure.java.io :as io])
(use 'clojure.stacktrace)

;; 「-cp」オプションが存在しない処理系への対処
(defn- -add-classpath []
    (let [libdir (io/file (.getParent (io/file *file*)) ".." "lib")
                urls [(io/as-url (io/file libdir "odb.jar"))
                      (io/as-url (io/file libdir "hsqldb.jar"))]
                loader (URLClassLoader. (into-array URL urls))]
        loader))

(defn -main []
    (let [info (doto (Properties.) (.setProperty "user" "sa")
                                   (.setProperty "password" ""))]
        (try
            (with-open [loader (-add-classpath)]
                (with-open [odbFile (-> (-> (Class/forName "com.k650250.odb.ODBFile" true loader) (.getMethod "open" (into-array [String]))) (.invoke [] (into-array ["sample.odb"])))]
                    (with-open [con (-> (-> (Class/forName "org.hsqldb.jdbcDriver" true loader) (.newInstance)) (.connect (.toUrl odbFile) info))]
                        (with-open [st (.createStatement con)]
                            (.println System/out "[更新前のデータ一覧]")
                            (let [sql (str (-> (-> odbFile (.getQuery "q_SAMPLE")) (.expand)))]
                                (with-open [records (.executeQuery st sql)]
                                    (while (.next records)
                                        (.println System/out (str (.getString records "key") "\t" (.getString records "value"))))))

                            ;; データ挿入
                            (.println System/out "追加データを入力してください。")
                            (let [value (.readLine (BufferedReader. (InputStreamReader. System/in (System/getProperty "native.encoding"))))]
                                (with-open [prep (.prepareStatement con "INSERT INTO \"t_sample\"(\"value\") VALUES(?)")]
                                    (.setString prep 1 value)
                                    (.executeUpdate prep)))

                            (.println System/out "[更新後のデータ一覧]")
                            (let [sql "SELECT * FROM \"t_sample\""]
                                (with-open [records (.executeQuery st sql)]
                                    (while (.next records)
                                        (.println System/out (str (.getString records "key") "\t" (.getString records "value"))))))))))

            (catch ClassNotFoundException e
                (.println System/err "[ERROR] ClassNotFoundException")
                (.println System/err "考えられる原因:")
                (.println System/err "- hsqldb.jar がクラス・パス指定されていない")
                (.println System/err "- odb.jar がクラス・ パス指定されていない")
                (.println System/err (str "Stack Trace: " (.getMessage e)) (print-stack-trace e)))
            (catch SQLException e
                (.println System/err "[ERROR] SQLException")
                (.println System/err "考えられる原因:")
                (.println System/err "- hsqldb.jar が正常に読み込まれていない")
                (.println System/err "- このデータベースを別のプロセスが使用中")
                (.println System/err "- 問合せ文中にクエリー名が含まれる")
                (.println System/err (str "Stack Trace: " (.getMessage e)) (print-stack-trace e))))))

(-main)
